lib-injection/sitecustomize.py

```diff
@@ -2,65 +2,238 @@
+        "tags": tags,
+    }
 
-debug_mode = os.environ.get("DD_TRACE_DEBUG", "").lower() in ("true", "1", "t")
-# Python versions that are supported by the current ddtrace release
-installable_py_versions = ("3.7", "3.8", "3.9", "3.10", "3.11", "3.12")
+
+def gen_telemetry_payload(telemetry_events):
+    return {
+        "metadata": {
+            "language_name": "python",
+            "language_version": PYTHON_VERSION,
+            "runtime_name": PYTHON_RUNTIME,
+            "runtime_version": PYTHON_VERSION,
+            "tracer_name": "python",
+            "tracer_version": INSTALLED_PACKAGES.get("ddtrace", "unknown"),
+            "pid": os.getpid(),
+        },
+        "points": telemetry_events,
+    }
+
+
+def send_telemetry(event):
+    event_json = json.dumps(event)
+    if not FORWARDER_EXECUTABLE or not TELEMETRY_ENABLED:
+        return
+    p = subprocess.Popen(
+        [FORWARDER_EXECUTABLE, str(os.getpid())],
+        stdin=subprocess.PIPE,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        universal_newlines=True,
+    )
+    p.stdin.write(event_json)
+    p.stdin.close()
 
 
 def _get_clib():
     """Return the C library used by the system.
 
     If GNU is not detected then returns MUSL.
     """
-    import platform
 
     libc, version = platform.libc_ver()
     if libc == "glibc":
         return "gnu"
     return "musl"
 
 
-def _log(msg, *args, level="info"):
+def _log(msg, *args, **kwargs):
     """Log a message to stderr.
 
     This function is provided instead of built-in Python logging since we can't rely on any logger
     being configured.
     """
-    if debug_mode:
+    level = kwargs.get("level", "info")
+    if DEBUG_MODE:
         asctime = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime())
         msg = "[%s] [%s] datadog.autoinstrumentation(pid: %d): " % (asctime, level.upper(), os.getpid()) + msg % args
         print(msg, file=sys.stderr)
 
 
+def runtime_version_is_supported(python_runtime, python_version):
+    supported_versions = RUNTIMES_ALLOW_LIST.get(python_runtime, {})
+    if not supported_versions:
+        return False
+    return (
+        supported_versions["min"].version <= parse_version(python_version).version <= supported_versions["max"].version
+    )
+
+
+def package_is_compatible(package_name, package_version):
+    installed_version = parse_version(package_version)
+    supported_version_spec = PKGS_ALLOW_LIST.get(package_name, Version((0,), ""))
+    if supported_version_spec.constraint in ("<", "<="):
+        return True  # minimum "less than" means there is no minimum
+    return installed_version.version >= supported_version_spec.version
+
+
 def _inject():
+    global INSTALLED_PACKAGES
+    global PYTHON_VERSION
+    global PYTHON_RUNTIME
+    global PKGS_ALLOW_LIST
+    INSTALLED_PACKAGES = build_installed_pkgs()
+    PYTHON_RUNTIME = platform.python_implementation().lower()
+    PYTHON_VERSION = ".".join(str(i) for i in sys.version_info[:2])
```

Comment:
```suggestion
    PYTHON_VERSION = platform.python_version()
```

Reply:
for Python runtime and version, any reason to generate them here in `_inject` vs in the global scope where they are defined?

Reply:
I put all of the nontrivial logic in this file into `_inject` to enable the small `try` block around the `_inject` call, catching all possible exceptions.